armi/reactor/blueprints/componentBlueprint.py

```diff
@@ -225,8 +225,52 @@ class can then use to modify the isotopics as necessary.
             constructedObject = components.factory(shape, [], kwargs)
             _setComponentFlags(constructedObject, self.flags, blueprint)
             insertDepletableNuclideKeys(constructedObject, blueprint)
+
+        # set the custom density for non-custom material components after construction
+        self.setCustomDensity(constructedObject, blueprint, matMods)
+
         return constructedObject
 
+    def setCustomDensity(self, constructedComponent, blueprint, matMods):
+        """Apply a custom density to a material with custom isotopics but not a 'custom material'."""
+        if self.isotopics is not None:
+            # must nest because one can't check for attribues on a NoneType
+            if blueprint.customIsotopics[self.isotopics].density > 0:
+                mat = materials.resolveMaterialClassByName(self.material)()
+                if not isinstance(mat, materials.Custom):
+                    # check for some problem cases
+                    if "TD_frac" in matMods.keys():
+                        runLog.warning(
+                            "Both TD_frac and a custom density (custom isotopics) has been specified for "
+                            "material {}. The custom density will override the density calculated using "
+                            "TD_frac.".format(self.material)
+                        )
+                    if not mat.density(Tc=self.Tinput) > 0:
+                        runLog.error(
+                            "A custom density has been assigned to material '{}', which has no baseline "
+                            "density. Only materials with a starting density may be assigned a density. "
+                            "This comes up e.g. if isotopics are assigned to 'Void'.".format(
+                                self.material
+                            )
+                        )
+                        raise ValueError(
+                            "Cannot apply custom densities to materials without density."
+                        )
+
+                    densityRatio = (
+                        blueprint.customIsotopics[self.isotopics].density
+                        / constructedComponent.density()
+                    )
+                    constructedComponent.changeNDensByFactor(densityRatio)
+
+                    runLog.important(
+                        "A custom material density was specified in the custom isotopics for non-custom "
+                        "material {} in component {}. The component density has been altered to "
+                        "{}.".format(
+                            mat, constructedComponent, constructedComponent.density()
+                        )
+                    )
```

Comment:
```suggestion
                    runLog.important(
                        "A custom material density was specified in the custom isotopics for non-custom "
                        "material {} in component {}. The component density has been altered to "
                        "{}.".format(
                            mat, constructedComponent, constructedComponent.density()
                        ),
                        single=True
                    )
```

Reply:
What I am suggesting here is that we make sure this log message is only ever printed once. Imagine if this message were printed hundreds of times for a single reactor, during construction. Boring, right?  This `single=True` should make sure no logging duplication happens.

Reply:
```suggestion
                    runLog.important(
                        "A custom material density was specified in the custom isotopics for non-custom "
                        "material {}. The component density has been altered to "
                        "{}.".format(
                            mat, constructedComponent.density()
                        ),
                        single=True
                    )
```
Good point, edited the suggestion to remove the component print to make the messages non-unique.